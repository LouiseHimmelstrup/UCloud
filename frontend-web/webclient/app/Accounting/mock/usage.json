{
    "data": {
        "usage": 1333333337,
        "dataType": "bytes",
        "quota": 13333333337,
        "title": "Storage Used"
    }
}