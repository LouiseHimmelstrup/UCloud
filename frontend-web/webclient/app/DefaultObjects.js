import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};