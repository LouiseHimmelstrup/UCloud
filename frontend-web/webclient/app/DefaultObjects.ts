import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState} from "Files";
import {SortOrder, SortBy} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Publication} from "Zenodo";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import * as SnackbarRedux from "Snackbar/Redux";
import * as FavoritesRedux from "Favorites/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum AnalysesStatusMap {
    "PENDING",
    "IN PROGRESS",
    "COMPLETED"
}

export enum RightsNameMap {
    "NONE" = "None",
    "READ" = "Read",
    "READ_WRITE" = "Read/Write"
}

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

function getFilesSortingColumnOrDefault(columnIndex: 0 | 1): SortBy {
    const sortingColumn = window.localStorage.getItem(`filesSorting${columnIndex}`);
    if (sortingColumn && Object.values(SortBy).includes(sortingColumn)) return sortingColumn as SortBy;
    switch (columnIndex) {
        case 0:
            window.localStorage.setItem("filesSorting0", SortBy.MODIFIED_AT);
            return SortBy.MODIFIED_AT;
        case 1:
            window.localStorage.setItem("filesSorting1", SortBy.SIZE);
            return SortBy.SIZE;
    }
}

function getItemOrDefault<T, T2>(itemName: string, defaultValue: T, en: T2): T {
    const item = window.localStorage.getItem(itemName);
    if (item && Object.values(en).includes(item)) {
        return item as unknown as T;
    }
    return defaultValue;
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FilesReduxObject extends ComponentWithPage<File> {
    sortOrder: SortOrder
    sortBy: SortBy
    path: string
    filesInfoPath: string
    fileSelectorError?: string
    sortingColumns: [SortBy, SortBy]
    fileSelectorLoading: boolean
    fileSelectorShown: boolean
    fileSelectorPage: Page<File>
    fileSelectorPath: string
    fileSelectorCallback: Function
    fileSelectorIsFavorites: boolean
    disallowedPaths: string[]
    invalidPath: boolean
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface ZenodoReduxObject extends ComponentWithPage<Publication> {
    connected: boolean
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    files?: Reducer<FilesReduxObject>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    zenodo?: Reducer<ZenodoReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
    detailedResult?: Reducer<DetailedResultReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

export type DetailedResultReduxObject = ComponentWithPage<File>

export const initDetailedResult = (): DetailedResultReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined
});

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    files: FilesReduxObject,
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    zenodo: ZenodoReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    detailedResult: DetailedResultReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects &
    FavoritesRedux.Objects &
    SnackbarRedux.Wrapper;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        files: initFiles(homeFolder),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        zenodo: initZenodo(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        detailedResult: initDetailedResult(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        ...FavoritesRedux.init(),
        ...SnackbarRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.ASCENDING
});

export const initZenodo = (): ZenodoReduxObject => ({
    connected: false,
    loading: false,
    page: emptyPage
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFiles = (homeFolder: string): FilesReduxObject => ({
    page: emptyPage,
    sortOrder: getItemOrDefault("sortOrder", SortOrder.ASCENDING, SortOrder),
    sortBy: getItemOrDefault("sortBy", SortBy.PATH, SortBy),
    loading: false,
    error: undefined,
    path: "",
    filesInfoPath: "",
    sortingColumns: [getFilesSortingColumnOrDefault(0), getFilesSortingColumnOrDefault(1)],
    fileSelectorLoading: false,
    fileSelectorShown: false,
    fileSelectorPage: emptyPage,
    fileSelectorPath: homeFolder,
    fileSelectorIsFavorites: false,
    fileSelectorCallback: () => undefined,
    fileSelectorError: undefined,
    disallowedPaths: [],
    invalidPath: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
