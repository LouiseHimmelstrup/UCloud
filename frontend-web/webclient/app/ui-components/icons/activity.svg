<svg viewBox="0 0 24 26" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linejoin="round" stroke-miterlimit="1.414"><path d="M21.602 0l-4.051 6.261h-5.29c-.486.056-.783.384-.783.782 0 .433.351.783.783.783h4.278l-1.013 1.565h-3.265c-.486.056-.783.385-.783.783 0 .432.351.783.783.783h2.252L13.5 12.522h-1.239c-.486.056-.783.384-.783.782 0 .433.351.783.783.783h.226L5.398 25.044H3.652a1.566 1.566 0 0 1-1.565-1.566V20.87H3.13a2.087 2.087 0 1 0 0-4.174H2.087v-2.087H3.13a2.087 2.087 0 0 0 0-4.174H2.087V8.348H3.13a2.087 2.087 0 1 0 0-4.174H2.087V1.565C2.087.701 2.788 0 3.652 0h17.95z" fill="#53657d"/><path d="M21.602 0h.833C23.299 0 24 .701 24 1.565v21.913c0 .808-.612 1.473-1.565 1.566H5.398l7.089-10.957h7.6c.432 0 .783-.35.783-.783 0-.398-.298-.726-.783-.782H13.5l1.013-1.565h5.574a.784.784 0 0 0 .783-.783c0-.398-.298-.727-.783-.783h-4.561l1.013-1.565h3.548c.432 0 .783-.35.783-.783 0-.398-.298-.726-.783-.782h-2.536L21.602 0zM3.13 19.826c.631-.056 1.044-.501 1.044-1.043 0-.576-.467-1.044-1.044-1.044H1.043a1.044 1.044 0 0 0 0 2.087H3.13zM3.13 13.565c.631-.056 1.044-.501 1.044-1.043 0-.576-.467-1.043-1.044-1.043H1.043a1.043 1.043 0 0 0 0 2.086H3.13zM3.13 7.305c.631-.056 1.044-.502 1.044-1.044 0-.576-.467-1.043-1.044-1.043H1.043a1.043 1.043 0 0 0 0 2.087H3.13z" fill="#8393a7"/></svg>