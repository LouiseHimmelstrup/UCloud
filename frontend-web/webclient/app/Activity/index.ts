import { ActivityReduxObject Â } from "DefaultObjects";

export type Activity = CountedActivity | TrackedActivity;

export type CountedOperations = "FAVORITE" | "DOWNLOAD";
export type TrackedOperations = "CREATE" | "UPDATE" | "DELETE" | "MOVED";

interface CountedActivityEntry {
    id: string
    path: string | null
    count: number
}

export interface CountedActivity {
    type: "counted"
    operation: CountedOperations
    entries: CountedActivityEntry[]
    timestamp: number
}

interface TrackedActivityFile {
    id: string
    path: string
}

export interface TrackedActivity {
    type: "tracked"
    operation: TrackedOperations
    files: TrackedActivityFile[]
    timestamp: number
}

export interface ActivityDispatchProps {
    fetchActivity: (pageNumber: number, pageSize: number) => void
    setError: (error?: string) => void
    setPageTitle: () => void
}

export interface ActivityProps extends ActivityReduxObject, ActivityDispatchProps { }