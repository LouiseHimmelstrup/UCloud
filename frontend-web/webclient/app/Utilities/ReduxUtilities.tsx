import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    dashboard,
    header,
    status,
    sidebar,
    notifications,
    simpleSearch,
    hookStore,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
