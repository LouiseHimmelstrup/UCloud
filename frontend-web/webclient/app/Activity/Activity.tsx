import * as React from "react";
import { connect } from "react-redux";
import { GroupedActivity, ActivityProps, ActivityDispatchProps } from "Activity";
import * as Module from "Activity";
import * as Pagination from "Pagination";
import * as moment from "moment";
import { getFilenameFromPath, replaceHomeFolder } from "Utilities/FileUtilities";
import { ActivityReduxObject, ReduxObject } from "DefaultObjects";
import { fetchActivity, setErrorMessage, setLoading } from "./Redux/ActivityActions";
import { updatePageTitle, setActivePage } from "Navigation/Redux/StatusActions";
import { Dispatch } from "redux";
import { fileInfoPage } from "Utilities/FileUtilities";
import * as Heading from "ui-components/Heading"
import Icon, { IconName } from "ui-components/Icon";
import { Flex, Text, Link } from "ui-components";
import Table, { TableRow, TableCell, TableBody, TableHeader, TableHeaderCell } from "ui-components/Table";
import { Dropdown, DropdownContent } from "ui-components/Dropdown";
import { Cloud } from "Authentication/SDUCloudObject";
import { MainContainer } from "MainContainer/MainContainer";
import styled from "styled-components";
import { SidebarPages } from "ui-components/Sidebar";
import { setRefreshFunction } from "Navigation/Redux/HeaderActions";
import { Spacer } from "ui-components/Spacer";

class Activity extends React.Component<ActivityProps> {

    public componentDidMount() {
        this.props.setPageTitle();
        this.props.setActivePage();
        this.props.fetchActivity(0, 100);
    }

    public componentWillUnmount() {
        this.props.clearRefresh();
    }

    public render() {
        const { fetchActivity, page, error, setError, loading, groupedEntries } = this.props;

        const main = (
            <React.StrictMode>
                <Pagination.List
                    loading={loading}
                    errorMessage={error}
                    onErrorDismiss={setError}
                    pageRenderer={() => <ActivityFeedGrouped activity={groupedEntries ? groupedEntries : []} />}
                    page={page}
                    onPageChanged={pageNumber => fetchActivity(pageNumber, page.itemsPerPage)}
                />
            </React.StrictMode>
        );

        const header = (<Spacer left={<Heading.h2>File Activity</Heading.h2>} right={
            <Pagination.EntriesPerPageSelector
                onChange={itemsPerPage => fetchActivity(page.pageNumber, itemsPerPage)}
                content="Activity per page"
                entriesPerPage={page.itemsPerPage}
            />
        } />);

        return (
            <MainContainer
                main={main}
                header={header}
            />
        );
    }
}


const ActivityFeedGrouped = ({ activity }: { activity: GroupedActivity[] }) => activity.length ? (
    <Table>
        <TableHeader>
            <TFRow>
                <TableHeaderCell width="7em" />
                <TableHeaderCell width="10.5em" />
                <TableHeaderCell width="99%" />
            </TFRow>
        </TableHeader>
        <TableBody>
            {activity.map((a, i) => <TrackedFeedActivity key={i} activity={a} />)}
        </TableBody>
    </Table>
) : null;

export const ActivityFeed = ({ activity }: { activity: Module.Activity[] }) =>
    <ActivityFeedGrouped activity={groupActivity(activity)!} />;

const ActivityOperation: React.FunctionComponent<{ event: Module.Activity }> = props => {
    switch (props.event.type) {
        case Module.ActivityType.MOVED: {
            return <>
                was moved to
                {" "}
                <b><Link to={fileInfoPage((props.event as Module.MovedActivity).newName)}>
                    {replaceHomeFolder((props.event as Module.MovedActivity).newName, Cloud.homeFolder)}
                </Link></b>
            </>;
        }

        case Module.ActivityType.FAVORITE: {
            const isFavorite = (props.event as Module.FavoriteActivity).favorite;
            if (isFavorite) {
                return <>was <b>added to favorites</b></>;
            } else {
                return <>was <b>removed from favorites</b></>;
            }
        }

        default: {
            return <>was <b>{operationToPastTense(props.event.type)}</b></>;
        }
    }
};

const ActivityEvent: React.FunctionComponent<{ event: Module.Activity }> = props => (
    <Text fontSize={1}>
        <b>
            <Link to={fileInfoPage(props.event.originalFilePath)}>
                {getFilenameFromPath(props.event.originalFilePath)}
            </Link>
        </b> <ActivityOperation event={props.event} />
    </Text>

);

const TrackedFeedActivity = ({ activity }: { activity: GroupedActivity }) => (
    <TFRow>
        <TableCell>
            <Dropdown>
                <Text fontSize={1} color="text">{moment(activity.timestamp).fromNow()}</Text>
                <DropdownContent>
                    {moment(activity.timestamp).format("llll")}
                </DropdownContent>
            </Dropdown>
        </TableCell>
        <TableCell>
            <Flex>
                <Icon mr="0.5em" name={eventIcon(activity.type).icon} />
                <Text fontSize={2}>{`Files ${operationToPastTense(activity.type)}`}</Text>
            </Flex>
        </TableCell>
        <TableCell>
            {activity.entries.map((item, idx) =>
                <ActivityEvent key={idx} event={item} />
            )}
        </TableCell>
    </TFRow>
);

const operationToPastTense = (operation: Module.ActivityType): string => {
    switch (operation) {
        case Module.ActivityType.DELETED: return "deleted";
        case Module.ActivityType.DOWNLOAD: return "downloaded";
        case Module.ActivityType.FAVORITE: return "favorited";
        case Module.ActivityType.INSPECTED: return "inspected";
        case Module.ActivityType.MOVED: return "moved";
        case Module.ActivityType.UPDATED: return "updated";
    }
}

interface EventIconAndColor {
    icon: IconName,
    color: "blue" | "green" | "red",
    rotation?: 45
}

const eventIcon = (operation: Module.ActivityType): EventIconAndColor => {
    switch (operation) {
        case Module.ActivityType.FAVORITE:
            return { icon: "starFilled", color: "blue" };
        case Module.ActivityType.DOWNLOAD:
            return { icon: "download", color: "blue" };
        case Module.ActivityType.UPDATED:
            return { icon: "refresh", color: "green" };
        case Module.ActivityType.DELETED:
            return { icon: "close", color: "red" };
        case Module.ActivityType.MOVED:
            return { icon: "move", color: "green" };
        default:
            return { icon: "ellipsis", color: "blue" }
    }
}

function groupActivity(items: Module.Activity[] = []): GroupedActivity[] {
    const result: GroupedActivity[] = [];
    let currentGroup: GroupedActivity | null = null;

    const pushGroup = () => {
        if (currentGroup != null) {
            result.push(currentGroup);
            currentGroup = null;
        }
    };

    const initializeGroup = (item: Module.Activity) => {
        currentGroup = {
            type: item.type,
            timestamp: item.timestamp,
            entries: [item]
        };
    };

    items.forEach(item => {
        if (currentGroup === null) {
            initializeGroup(item);
        } else {
            if (currentGroup.type !== item.type || Math.abs(item.timestamp - currentGroup.timestamp) > (1000 * 60 * 15)) {
                pushGroup();
                initializeGroup(item);
            } else {
                currentGroup.entries.push(item);
            }
        }
    });

    pushGroup();
    return result;
}

const TFRow = styled(TableRow)`
    vertical-align: top;
`;

const mapStateToProps = ({ activity }: ReduxObject): ActivityReduxObject & Module.ActivityOwnProps => ({
    ...activity,
    groupedEntries: groupActivity(activity.page.items)
});

const mapDispatchToProps = (dispatch: Dispatch): ActivityDispatchProps => ({
    fetchActivity: (pageNumber, pageSize) => {
        const fetch = async () => {
            dispatch(setLoading(true));
            dispatch(await fetchActivity(pageNumber, pageSize));
        };
        fetch();
        dispatch(setRefreshFunction(fetch))
    },
    setError: error => dispatch(setErrorMessage(error)),
    setPageTitle: () => dispatch(updatePageTitle("Activity")),
    setActivePage: () => dispatch(setActivePage(SidebarPages.Activity)),
    clearRefresh: () => dispatch(setRefreshFunction())
});

export default connect(mapStateToProps, mapDispatchToProps)(Activity);