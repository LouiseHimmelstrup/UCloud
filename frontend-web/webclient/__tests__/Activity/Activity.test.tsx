import * as React from "react";
import Activity, { ActivityFeed } from "Activity/Activity";
import { create } from "react-test-renderer";
import { initActivity } from "DefaultObjects";
import activity from "Activity/Redux/ActivityReducer";
import { configureStore } from "Utilities/ReduxUtilities";
import { Provider } from "react-redux";
import { activityPage } from "../mock/Activity";
import { MemoryRouter } from "react-router";

describe("Activity", () => {
    test("Mount", () => {
        const store = configureStore({ activity: initActivity() }, { activity })
        expect(create(<Provider store={store}><Activity /></Provider>).toJSON()).toMatchSnapshot();
    });
});

describe("Activity Feed", () => {
    test("Mount, no activity", () => {
        expect(create(<ActivityFeed activity={[]} />))
    });

    test("Mount, with activity", () => {
        expect(create(<MemoryRouter><ActivityFeed activity={activityPage.items} /></MemoryRouter>))
    });
})