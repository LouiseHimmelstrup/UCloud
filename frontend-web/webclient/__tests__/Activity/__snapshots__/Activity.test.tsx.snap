// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity Feed Mount, no activity 1`] = `null`;

exports[`Activity Feed Mount, with activity 1`] = `
<div
  className="ui feed"
>
  <div
    className="event"
  >
    <div
      className="label"
    >
      <i
        aria-hidden="true"
        className="move icon"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="date"
      >
        a month ago
      </div>
      <div
        className="summary"
      >
        Files moved
      </div>
      <div
        className="text extra"
      >
        <div>
          <b>
            <a
              href="/fileInfo//home/jonas@hinchely.dk/ABC.jpg"
              onClick={[Function]}
            >
              ABC.jpg
            </a>
          </b>
           was 
          <b>
            moved
          </b>
        </div>
      </div>
    </div>
  </div>
  <div
    className="event"
  >
    <div
      className="label"
    >
      <i
        aria-hidden="true"
        className="star outline icon"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="date"
      >
        a month ago
      </div>
      <div
        className="summary"
      >
        Files unfavorited
      </div>
      <div
        className="text extra"
      >
        <div>
          <b>
            <a
              href="/fileInfo//home/jonas@hinchely.dk/ABC.jpg"
              onClick={[Function]}
            >
              ABC.jpg
            </a>
          </b>
           was 
          <b>
            unfavorited
          </b>
        </div>
      </div>
    </div>
  </div>
  <div
    className="event"
  >
    <div
      className="label"
    >
      <i
        aria-hidden="true"
        className="delete icon"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="date"
      >
        a month ago
      </div>
      <div
        className="summary"
      >
        Files deleted
      </div>
      <div
        className="text extra"
      />
    </div>
  </div>
  <div
    className="event"
  >
    <div
      className="label"
    >
      <i
        aria-hidden="true"
        className="star icon"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="date"
      >
        a month ago
      </div>
      <div
        className="summary"
      >
        Files favorited
      </div>
      <div
        className="text extra"
      >
        <div>
          <b>
            <a
              href="/fileInfo//home/jonas@hinchely.dk/ABC.jpg"
              onClick={[Function]}
            >
              ABC.jpg
            </a>
          </b>
           was 
          <b>
            favorited
          </b>
        </div>
      </div>
    </div>
  </div>
  <div
    className="event"
  >
    <div
      className="label"
    >
      <i
        aria-hidden="true"
        className="refresh icon"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="date"
      >
        a month ago
      </div>
      <div
        className="summary"
      >
        Files updated
      </div>
      <div
        className="text extra"
      />
    </div>
  </div>
  <div
    className="event"
  >
    <div
      className="label"
    >
      <i
        aria-hidden="true"
        className="refresh icon"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="date"
      >
        2 months ago
      </div>
      <div
        className="summary"
      >
        Files updated
      </div>
      <div
        className="text extra"
      >
        <div>
          <b>
            <a
              href="/fileInfo//home/jonas@hinchely.dk/ABC.jpg"
              onClick={[Function]}
            >
              ABC.jpg
            </a>
          </b>
           was 
          <b>
            updated
          </b>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Activity Mount 1`] = `
Array [
  <h1
    className="sc-gzVnrw icdHnu"
    cursor="inherit"
    fontSize={6}
  >
    File Activity
  </h1>,
  <div
    className="sc-bZQynM cNFsWw"
  >
    <div
      className="sc-htpNat hjOqln"
    />
    <div
      className="sc-cvbbAY iiWipZ"
    >
      <div
        className="sc-bdVaJa hfvaUK"
      >
        <span
          className="sc-gzVnrw hNFoCO"
          cursor="pointer"
          onClick={[Function]}
        >
          <div
            className="sc-bZQynM bqIUrb"
          >
            <div
              className="sc-htpNat llOEfe"
            >
               
              Items per page 25
            </div>
            <div
              className="sc-htpNat llOEfe"
            >
              <svg
                className="sc-dnqmqq fuFxzC"
                cursor="auto"
                fill="currentcolor"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M7.4 8l4.6 4.6L16.6 8 18 9.4l-6 6-6-6L7.4 8z"
                />
              </svg>
            </div>
          </div>
        </span>
      </div>
      <i
        className="fas fa-sync fa-spin"
        onClick={[Function]}
      />
    </div>
  </div>,
  <div
    className="sc-bZQynM bqIUrb"
  >
    <div
      className="sc-htpNat knGhlu"
      width="50%"
    />
    <i
      className="fas fa-circle-notch fa-spin"
    />
  </div>,
  <div />,
]
`;
